lib/common/operations.c

```diff
@@ -47,80 +47,106 @@ pcmk__op_key(const char *rsc_id, const char *op_type, guint interval_ms)
 gboolean
 parse_op_key(const char *key, char **rsc_id, char **op_type, guint *interval_ms)
 {
-    char *notify = NULL;
-    char *mutable_key = NULL;
-    char *mutable_key_ptr = NULL;
     size_t len = 0, offset = 0;
-    unsigned long long ch = 0;
-    guint local_interval_ms = 0;
+    char *found = NULL;
 
     // Initialize output variables in case of early return
     if (rsc_id) {
         *rsc_id = NULL;
     }
+
     if (op_type) {
         *op_type = NULL;
     }
+
     if (interval_ms) {
         *interval_ms = 0;
     }
 
     CRM_CHECK(key && *key, return FALSE);
 
-    // Parse interval at end of string
     len = strlen(key);
     offset = len - 1;
+
+    // Parse interval at end of string
     while ((offset > 0) && isdigit(key[offset])) {
-        ch = key[offset] - '0';
-        for (int digits = len - offset; digits > 1; --digits) {
-            ch = ch * 10;
-        }
-        local_interval_ms += ch;
         offset--;
     }
+
     if (interval_ms) {
-        *interval_ms = local_interval_ms;
-    }
+        unsigned long long l;
 
-    CRM_CHECK((offset != (len - 1)) && (key[offset] == '_'), return FALSE);
+        errno = 0;
+        l = strtoul(&(key[offset+1]), NULL, 10);
 
-    mutable_key = strndup(key, offset);
-    offset--;
+        if (errno != 0) {
+            return FALSE;
+        }
 
-    while (offset > 0 && key[offset] != '_') {
-        offset--;
+        *interval_ms = (guint) l;
     }
 
-    CRM_CHECK(key[offset] == '_',
-              free(mutable_key); return FALSE);
-
-    mutable_key_ptr = mutable_key + offset + 1;
+    // Verify we're at the separator between the operation and interval.
+    if (offset == len-1 || key[offset] != '_') {
+        if (interval_ms) {
+            *interval_ms = 0;
+        }
 
-    if (op_type) {
-        *op_type = strdup(mutable_key_ptr);
+        return FALSE;
     }
 
-    mutable_key[offset] = 0;
-    offset--;
+    // Try to find an operation type with an underscore in its name.  Valid
+    // operations are migrate_from, migrate_to, pre_notify_*, and post_notify_*.
+    if ((found = strcasestr(key, "_migrate_from")) != NULL ||
+        (found = strcasestr(key, "_migrate_to")) != NULL ||
+        (found = strcasestr(key, "_pre_notify")) != NULL ||
+        (found = strcasestr(key, "_post_notify")) != NULL) {
```

Comment:
I've been playing around with this regex for a bit, and I'm not sure it's going to work.  For an input string where there's underscores in the operation type, the first match is going to greedily grab characters through the operation type as well.  For example, `imagestoreclone:1_post_notify_stop_0` is getting split up as `imagestoreclone:1_post_notify`, `stop`, and `0`.

I don't know how to modify the regex aside from enabling lazy matching, which only exists in pcre and not in posix.

Reply:
Ah bummer. I think it's because "stop" matches the first [a-z]+ so it doesn't need to check the other possibilities. Probably the only way around that would be two regexes, one with the notify/migrate matches and one with [a-z]+ (or only using a regex for the notify/migrate cases). (Or enumerating all possible actions but that would be ugly, brittle, and inefficient.)

Reply:
Switching from the code on master to the two regex solution is not encouraging:
```
clumens@localhost:~/src/pacemaker$ tools/crm_simulate --repeat 10000 --profile /tmp/blah
* Testing /tmp/blah/params-6.xml ... 1217.44 secs

clumens@localhost:~/src/pacemaker$ tools/crm_simulate --repeat 10000 --profile /tmp/blah
* Testing /tmp/blah/params-6.xml ... 1478.69 secs
```

I don't have numbers for my original non-regex based code, but I can generate them in the morning.

Reply:
The non-regex code comes in halfway between, of course:
```
clumens@localhost:~/src/pacemaker$ tools/crm_simulate --repeat 10000 --profile /tmp/blah
* Testing /tmp/blah/params-6.xml ... 1301.31 secs
```

Reply:
Yeah I was hoping that option would be enlightening but there's a ton of variation even on a machine that's not doing much else. BTW that's 10,000 repetitions so each one is 0.13 seconds, not horrible.

Reply:
Okay, running it on boomstick:
```
master:

[clumens@boomstick pacemaker]$ PCMK_schema_directory=xml tools/crm_simulate --repeat 10000 --profile /tmp/blah
* Testing /tmp/blah/params-6.xml ... 1357.52 secs


rewritten non-regex code:

[clumens@boomstick pacemaker]$ PCMK_schema_directory=xml tools/crm_simulate --repeat 10000 --profile /tmp/blah
* Testing /tmp/blah/params-6.xml ... 1352.55 secs


two regexes:

[clumens@boomstick pacemaker]$ PCMK_schema_directory=xml tools/crm_simulate --repeat 10000 --profile /tmp/blah
* Testing /tmp/blah/params-6.xml ... 1675.89 secs
```

Personally, I prefer the one regex version but I don't see a way of making that work, aside from introducing a pcre dependency.  The two regex version is likely always going to be slower because some percentage of runs are going to require testing both regexes.  And it's worse because I think we have to test for the migrate/notify case first, which is probably also the rarer of the two.  My hand crafted replacement code seems to be on par with the original and seems to be compatible with that version's shortcomings (the regexes handle those commented out tests), but is also a lot of hand written string processing code.

Reply:
For the hybrid method:
```
[clumens@boomstick pacemaker]$ PCMK_schema_directory=xml tools/crm_simulate --repeat 10000 --profile /tmp/blah
* Testing /tmp/blah/params-6.xml ... 1431.81 secs
```

That seems acceptable to me.  I'll get the patches finished up and update this PR, and maybe we can find some extra speed savings.